wave_geometry/include/wave/geometry/numerical_test_functors.hpp

```diff
@@ -33,26 +33,122 @@ class RotateAndJacobianJparamFunctor {
+    RotateAndJacobianJparamFunctor(const Rotation &input_rotation,
+                                   const Vec3 &input_point) {
         this->R = input_rotation;
         this->P = input_point;
     }
 
-    Vec3 operator()(Vec3 input_point) {
+    Vec3 operator()(const Vec3 &input_point) {
         Rotation Rp = this->R;
         Rp = Rp.manifoldPlus(input_point);
         Vec3 output_point = Rp.rotate(P);
         return output_point;
     }
 };
 
+class ComposeAndJacobianJLeftFunctor {
+ public:
+    Rotation R_left;
+    Rotation R_right;
+    ComposeAndJacobianJLeftFunctor(const Rotation &R_left,
+                                   const Rotation &R_right) {
+        this->R_left = R_left;
+        this->R_right = R_right;
+    }
 
-// Implement the numerical differentiation, following procedure
-// for a 5 point stencil https://en.wikipedia.org/wiki/Five-point_stencil
-// Exact step size representation is outlined in
-// "Numerical Recipes by Brian P. Flannery,
-// Saul Teukolsky, and William H. Press, chapter 5.7.
+    Rotation operator()(const Vec3 &perturbation) {
+        Mat3 J;
+        Rotation R_perturbed = this->R_left;
+        R_perturbed = R_perturbed.manifoldPlus(perturbation);
+        return R_perturbed.composeAndJacobian(R_right, J, J);
+    }
+};
 
+class ComposeAndJacobianJRightFunctor {
+ public:
+    Rotation R_left;
+    Rotation R_right;
+    ComposeAndJacobianJRightFunctor(const Rotation &R_left,
+                                    const Rotation &R_right) {
+        this->R_left = R_left;
+        this->R_right = R_right;
+    }
+
+    Rotation operator()(const Vec3 &perturbation) {
+        Mat3 J;
+        Rotation R_perturbed = this->R_right;
+        R_perturbed = R_perturbed.manifoldPlus(perturbation);
+        return R_left.composeAndJacobian(R_perturbed, J, J);
+    }
+};
+
+class InverseAndJacobianFunctor {
+ public:
+    Rotation R;
+    InverseAndJacobianFunctor(const Rotation &R) {
+        this->R = R;
+    }
+
+    Rotation operator()(const Vec3 &perturbation) {
+        Mat3 J;
+        Rotation R_perturbed = this->R;
+        R_perturbed = R_perturbed.manifoldPlus(perturbation);
+        return R_perturbed.inverseAndJacobian(J);
+    }
+};
+
+class LogMapAndJacobianFunctor {
+ public:
+    Rotation R;
+    LogMapAndJacobianFunctor(const Rotation &R) {
+        this->R = R;
+    }
+
+    Vec3 operator()(const Vec3 &perturbation) {
+        Mat3 J;
+        Rotation R_perturbed = this->R;
+        R_perturbed = R_perturbed.manifoldPlus(perturbation);
+        return Rotation::logMapAndJacobian(R_perturbed, J);
+    }
+};
+
+
+class ManifoldMinusAndJacobianJLeftFunctor {
+ public:
+    Rotation R_left;
+    Rotation R_right;
+    ManifoldMinusAndJacobianJLeftFunctor(const Rotation &R_left,
+                                         const Rotation &R_right) {
+        this->R_left = R_left;
+        this->R_right = R_right;
+    }
```

Comment:
No big deal but in future prefer defining constructors via list initialization: https://stackoverflow.com/questions/926752/why-should-i-prefer-to-use-member-initialization-list